### PR TITLE
GPTBRIDGE-19: Paginate List Assistants

### PR DESCRIPTION
**Pull request by froggomad**
# Story
[GPTBRIDGE-19](https://infinitiqtech.atlassian.net/browse/GPTBRDGE-19)

# Description
Adds pagination to the `listAssistants` method

# Checklist
**Required**
- [ ] Implemented a unit test
- [x] Modified a unit test
- [ ] Didn't require unit testing

**Optional**
- [ ] Added a 3rd-party dependency
- [x] Modified the public API
  - Modified listAssistants method to include pagination

## Summary by CodeRabbit

- **New Features**
  - Added dynamic fetching and display of assistants in a paginated manner in the Assistant List View.
  - Introduced pagination parameters to the assistant listing functionality.

- **Bug Fixes**
  - Improved error message logging for better clarity.

- **Tests**
  - Updated tests to include new pagination properties and assertions for `firstId`, `lastId`, and `hasMore`.

- **Refactor**
  - Enhanced endpoint URL construction to include query items dynamically.

### CODE REVIEW

File: My GPT Dashboard/My GPT Dashboard/AssistantListView.swift [hunk 1]
```diff
@@ -32,8 +26,27 @@ struct AssistantListView: View {
...
+                    let paginationRequest = PaginatedRequestParameters(
+                        limit: 10,
+                        startAfter: lastId ?? "",
+                        order: .descending
+                    )
+                    let response = try await GPTBridge.listAssistants(paginatedBy: paginationRequest)
+                    hasMore = response.hasMore
+                    assistants += response.data
+                }
+
```
[the remark this answers is not in the document]
Comment (froggomad) [reply My GPT Dashboard/My GPT Dashboard/AssistantListView.swift:47]: this is a sample project

File: Sources/GPTBridge/Models/ListAssistants.swift [hunk 2]
```diff
@@ -7,9 +7,97 @@
...
-    let data: [Assistant]
+public enum PaginationOrder: String, Encodable {
+    case ascending = "asc"
+    case createdAt = "created_at"
+    case descending = "desc"
+}
+/// Parameters for paginated requests
+public struct PaginatedRequestParameters {
+    /// A limit on the number of objects to be returned. Limit can range between 1 and 100, and the default is 20.
+    private(set) var limit: Int? = nil
```
[the remark this answers is not in the document]
Comment (froggomad) [reply Sources/GPTBridge/Models/ListAssistants.swift:18]: this is a necessary hack due to using a private setter method for shared init (can't reference self without initializing first)

File: Sources/GPTBridge/GPTBridge.swift [hunk 3]
```diff
@@ -125,54 +182,101 @@ public class GPTBridge {
...
+            loopStatus
+        ) {
+            let failedRunHandler = FailedRunHandler(
+                runThreadResponse: currentRunResponse
+            )
             try await failedRunHandler.handle()
-            return MessageRunStepResult(message: failedRunHandler.lastError.localizedString) // TODO: Anti-pattern. create FailedRunStepResult
+            return MessageRunStepResult(
+                message: failedRunHandler.lastError.localizedString
+            ) // TODO: Anti-pattern. create FailedRunStepResult
```
[the remark this answers is not in the document]
Comment (froggomad) [reply Sources/GPTBridge/GPTBridge.swift:248]: OOS

File: Sources/GPTBridge/GPTBridge.swift [hunk 4]
```diff
@@ -45,61 +45,115 @@ public class GPTBridge {
...
+            order: paginatedBy?.order
+        )
+        
+        return try await requestManager
+            .makeRequest(
+                endpoint: .listAssistants(limit: paginatedBy?.limit, order: paginatedBy?.order, before: paginatedBy?.startBefore, after: paginatedBy?.startAfter),
+                method: .GET,
+                requestData: listAssistantRequest
+            )
+    }
```
[the remark this answers is not in the document]
Comment (froggomad) [reply Sources/GPTBridge/GPTBridge.swift]: sure, let's see what ya got

File: Sources/GPTBridge/GPTBridge.swift [hunk 5]
```diff
@@ -45,43 +45,89 @@ public class GPTBridge {
...
+            limit: paginatedBy?.limit,
+            order: paginatedBy?.order
+        )
+
+        return try await requestManager
+            .makeRequest(
+                endpoint: .listAssistants(limit: paginatedBy?.limit, order: paginatedBy?.order, before: paginatedBy?.startBefore, after: paginatedBy?.startAfter),
+                method: .GET,
+                requestData: listAssistantRequest
+            )
```
[the remark this answers is not in the document]
Comment (froggomad) [reply Sources/GPTBridge/GPTBridge.swift:82]: you read that wrong...

[the remark this answers is not in the document]
Comment (froggomad) [reply Sources/GPTBridge/GPTBridge.swift:82]: your suggestion is already implemented. You must have misinterpreted the diff

File: Sources/GPTBridge/GPTBridge.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (froggomad) [reply Sources/GPTBridge/GPTBridge.swift]: 1. can't subclass a struct
2. that's not the right signature for the method

I made a protocol and spy and separated URL creation from request/response handling so I can test the paginated parameters in URL creation

File: Sources/GPTBridge/GPTBridge.swift [hunk 6]
```diff
@@ -188,31 +292,69 @@ public class GPTBridge {
...
+            runId: runId
+        )
     }
 
-    static func getMessageId(threadId: String, runId: String) async throws -> String { // TODO: Move to MessageRunHandler
-        let endpoint = AssistantEndpoint.getMessageId(threadId: threadId, runId: runId)
+    static func getMessageId(
+        threadId: String,
+        runId: String
+    ) async throws -> String { // TODO: Move to MessageRunHandler
```
[the remark this answers is not in the document]
Comment (froggomad) [reply Sources/GPTBridge/GPTBridge.swift:322]: OOS

**Comment by brookshanes** [comment]
Wow I don't even know what I'm doing here when coderabbitai is at the helm. Incredible!

**Comment by kdubroff** [comment]
> Wow I don't even know what I'm doing here when coderabbitai is at the helm. Incredible!

I value your opinion > 9000 over codeRabbitStew... that's what

**Comment by kdubroff** [comment]
Coderabbit is a tool @brookshanes is an MVP

**Comment by froggomad** [comment]
> Wow I don't even know what I'm doing here when coderabbitai is at the helm. Incredible!

I value your opinion > 9000 over coderabbit, that's what. Coderabbit is a tool @brookshanes is an MVP

**Pull request merged**